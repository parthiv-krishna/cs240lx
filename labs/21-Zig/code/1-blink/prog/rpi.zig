extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


