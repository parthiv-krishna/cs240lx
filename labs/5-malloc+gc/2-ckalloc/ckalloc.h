// weak test of malloc/free: checks that subsequent allocations are
// <sizeof (void *)> bytes from each other.

#ifndef __CKALLOC_H__
#define __CKALLOC_H__

#include "src-loc.h"

typedef enum {  ALLOCED = 11, FREED } state_t;

// pull the remainder into the second redzone.
typedef struct ck_hdr {
    struct ck_hdr *next;
    uint32_t nbytes_alloc;  // how much the user requested to allocate.
    uint32_t state;          // state of the block: { ALLOCED, FREED }

    src_loc_t alloc_loc;    // location they called ckalloc() at.

    // used for gc: i didn't cksum these.
    uint32_t refs_start;    // number of pointers to the start of the block.
    uint32_t refs_middle;   // number of pointers to the middle of the block.

    uint16_t mark;          // 0 initialize.
} hdr_t;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void);

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p);

// get the number of bytes for hdr <h>
static inline unsigned ck_nbytes(hdr_t *h) {
    return h->nbytes_alloc;
}
// get the start of allocated data for <h>
void *ck_hdr_start(hdr_t *h);
// get the end of allocated data for <h>
void *ck_hdr_end(hdr_t *h);

// is <ptr> in block <h>?
unsigned 
ck_ptr_in_block(hdr_t *h, void *ptr);

// is pointer <p> on the allocated list?
int ck_ptr_is_alloced(void *ptr);

#define ckalloc(_n) (ckalloc)(_n, SRC_LOC_MK())
#define ckfree(_ptr) (ckfree)(_ptr, SRC_LOC_MK())
void *(ckalloc)(uint32_t nbytes, src_loc_t loc);
void (ckfree)(void *addr, src_loc_t loc);

#endif
