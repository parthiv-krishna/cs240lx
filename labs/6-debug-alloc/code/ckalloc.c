// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing %p\n", addr);
    
    h->state = FREED;
    memset(h + 1, 0, h->nbytes_alloc);
    assert(ck_ptr_is_alloced(addr));
    kr_free(h);
    list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}

// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {

    hdr_t *h = kr_malloc(nbytes + sizeof *h);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;


    loc_debug(loc, "allocating %p\n", h);

    void *addr = (void *)(h + 1);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}
