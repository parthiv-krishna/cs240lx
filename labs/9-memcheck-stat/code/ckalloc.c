// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;
static hdr_t *free_list;

unsigned ck_verbose_p = 0;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}


void make_redzone(char *p, unsigned n) {
    memset(p, REDZONE_VAL, n);
}

int check_redzone(char *p, unsigned n) {
    for(unsigned i = 0; i < n; i++) {
        if(p[i] != REDZONE_VAL) {
            printk("found error at %p, %u, %d\n", p, i, p[i]);
            return 4*i + 1;
        }
    }
    return 0;
}

static int ckalloc_count = 1;

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    int ofst = check_redzone(h->rz1, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        
    
    ofst = check_redzone((char *)addr + h->nbytes_alloc, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1 + h->nbytes_alloc);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        

    

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing block=%u [addr=%p]\n", h->block_id, addr);
    
    h->state = FREED;
    make_redzone((char *)(h + 1), h->nbytes_alloc); // nuke data
    assert(ck_ptr_is_alloced(addr));

    list_remove(&alloc_list, h);
    h->next = free_list;
    free_list = h;


    // list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}


// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {


    hdr_t *h = kr_malloc(nbytes + sizeof *h + REDZONE_NBYTES);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;
    h->block_id = ckalloc_count++;

    loc_debug(loc, "allocating block=%u [addr=%p]\n", h->block_id, h);

    void *addr = (void *)(h + 1);

    make_redzone(h->rz1, REDZONE_NBYTES);
    make_redzone((char *)addr + nbytes, REDZONE_NBYTES);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}


int ck_heap_errors() {
    int errors = 0;
    int nblocks = 0;
    for(hdr_t *h = alloc_list; h; h = h->next) {
        nblocks++;

        if (h->state == ALLOCED) {
            if (ck_verbose_p) printk("alloc block block=%u [addr=[%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("free block %u [addr=%p] in alloc list", h->block_id, h);
        }
    }

    for(hdr_t *h = free_list; h; h = h->next) {
        nblocks++;

        if (h->state == FREED) {
            if (ck_verbose_p) printk("free block block=%u [addr=%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1), h->nbytes_alloc) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("alloc block %u [addr=%p] in free list", h->block_id, h);
        }

    }

    return errors;
}
