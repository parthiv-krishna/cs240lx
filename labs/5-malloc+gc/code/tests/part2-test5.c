// simple compaction.
#include "rpi.h"
#include "ckalloc.h"

#define N 10

// allocate a linked list.  do a leak detection.
void *test(void) {
    struct list {
        int x; 
        struct list *next;
        struct list *prev;
    } *h = 0;

    for(int i = 0; i < N; i++) {
        struct list *e = ckalloc(sizeof *h);
        memset(e, 0, sizeof *e);
        e->x = 0;
        if(h) {
            assert(!h->prev);
            h->prev = e;
            e->next = h;
        }
        h = e;
    }
    // compiler optimizes this stuff away.
    // h= 0;
    // even this:
    //  memset(&h, 0, sizeof h);
    // perhaps if we put it in another file?
    check_no_leak();
    // return to try to get around compiler tricks.
    return h;
}

// mild attempt to make sure that the random values from <test> are 
// gone by the time we get back to main.
void bar(void) {
    test();
}

void notmain(void) {
    printk("GC test: checking that a doubly-linked list is handled.\n");


    // note: if we do a check after this, may miss b/c the value is laying around
    // on the stack or reg
    bar();

    // should detect leaks.
    check_should_leak();
    // should compact everything.
    printk("compacted = %d\n", ck_gc());
    // it must be the case that after we compact we should not 
    // detect any leak.
    check_no_leak();
}
