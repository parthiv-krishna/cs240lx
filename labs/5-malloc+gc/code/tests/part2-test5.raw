find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/part2-test5.bin> about to boot
HASH: simple_boot: sending 10788 bytes, crc32=f5fe8594
waiting for a start
putting code
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
GC test: checking that a doubly-linked list is handled.
tests/part2-test5.c:test:16:allocating 0x17fff4
found 0x17fff4 in 0x17ffd0
tests/part2-test5.c:test:16:successful alloc of 0x17fff4
tests/part2-test5.c:test:16:allocating 0x17ffbc
found 0x17ffbc in 0x17ff98
tests/part2-test5.c:test:16:successful alloc of 0x17ffbc
tests/part2-test5.c:test:16:allocating 0x17ff84
found 0x17ff84 in 0x17ff60
tests/part2-test5.c:test:16:successful alloc of 0x17ff84
tests/part2-test5.c:test:16:allocating 0x17ff4c
found 0x17ff4c in 0x17ff28
tests/part2-test5.c:test:16:successful alloc of 0x17ff4c
tests/part2-test5.c:test:16:allocating 0x17ff14
found 0x17ff14 in 0x17fef0
tests/part2-test5.c:test:16:successful alloc of 0x17ff14
tests/part2-test5.c:test:16:allocating 0x17fedc
found 0x17fedc in 0x17feb8
tests/part2-test5.c:test:16:successful alloc of 0x17fedc
tests/part2-test5.c:test:16:allocating 0x17fea4
found 0x17fea4 in 0x17fe80
tests/part2-test5.c:test:16:successful alloc of 0x17fea4
tests/part2-test5.c:test:16:allocating 0x17fe6c
found 0x17fe6c in 0x17fe48
tests/part2-test5.c:test:16:successful alloc of 0x17fe6c
tests/part2-test5.c:test:16:allocating 0x17fe34
found 0x17fe34 in 0x17fe10
tests/part2-test5.c:test:16:successful alloc of 0x17fe34
tests/part2-test5.c:test:16:allocating 0x17fdfc
found 0x17fdfc in 0x17fdd8
tests/part2-test5.c:test:16:successful alloc of 0x17fdfc
found 0x17fdfc in 0x17fdd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fdfc
ck-gc.c:mark:101:CK_DEBUG:0x7ffff7c: referenced start = 0x17fdfc
found 0x17fe34 in 0x17fe10
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fe34
ck-gc.c:mark:101:CK_DEBUG:0x17fe00: referenced start = 0x17fe34
found 0x17fe6c in 0x17fe48
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fe6c
ck-gc.c:mark:101:CK_DEBUG:0x17fe38: referenced start = 0x17fe6c
found 0x17fea4 in 0x17fe80
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fea4
ck-gc.c:mark:101:CK_DEBUG:0x17fe70: referenced start = 0x17fea4
found 0x17fedc in 0x17feb8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fedc
ck-gc.c:mark:101:CK_DEBUG:0x17fea8: referenced start = 0x17fedc
found 0x17ff14 in 0x17fef0
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff14
ck-gc.c:mark:101:CK_DEBUG:0x17fee0: referenced start = 0x17ff14
found 0x17ff4c in 0x17ff28
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff4c
ck-gc.c:mark:101:CK_DEBUG:0x17ff18: referenced start = 0x17ff4c
found 0x17ff84 in 0x17ff60
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff84
ck-gc.c:mark:101:CK_DEBUG:0x17ff50: referenced start = 0x17ff84
found 0x17ffbc in 0x17ff98
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ffbc
ck-gc.c:mark:101:CK_DEBUG:0x17ff88: referenced start = 0x17ffbc
found 0x17fff4 in 0x17ffd0
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fff4
ck-gc.c:mark:101:CK_DEBUG:0x17ffc0: referenced start = 0x17fff4
found 0x17ffbc in 0x17ff98
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ffbc
ck-gc.c:mark:101:CK_DEBUG:0x17fffc: referenced start = 0x17ffbc
found 0x17ff84 in 0x17ff60
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff84
ck-gc.c:mark:101:CK_DEBUG:0x17ffc4: referenced start = 0x17ff84
found 0x17ff4c in 0x17ff28
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff4c
ck-gc.c:mark:101:CK_DEBUG:0x17ff8c: referenced start = 0x17ff4c
found 0x17ff14 in 0x17fef0
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17ff14
ck-gc.c:mark:101:CK_DEBUG:0x17ff54: referenced start = 0x17ff14
found 0x17fedc in 0x17feb8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fedc
ck-gc.c:mark:101:CK_DEBUG:0x17ff1c: referenced start = 0x17fedc
found 0x17fea4 in 0x17fe80
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fea4
ck-gc.c:mark:101:CK_DEBUG:0x17fee4: referenced start = 0x17fea4
found 0x17fe6c in 0x17fe48
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fe6c
ck-gc.c:mark:101:CK_DEBUG:0x17feac: referenced start = 0x17fe6c
found 0x17fe34 in 0x17fe10
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fe34
ck-gc.c:mark:101:CK_DEBUG:0x17fe74: referenced start = 0x17fe34
found 0x17fdfc in 0x17fdd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fdfc
ck-gc.c:mark:101:CK_DEBUG:0x17fe3c: referenced start = 0x17fdfc
ck-gc.c:mark_all:204:stack has 40 words
found 0x17fdfc in 0x17fdd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fdfc
ck-gc.c:mark:101:CK_DEBUG:0x7ffff7c: referenced start = 0x17fdfc
found 0x17fdfc in 0x17fdd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fdfc
ck-gc.c:mark:101:CK_DEBUG:0x7ffffb0: referenced start = 0x17fdfc
found 0x17fdfc in 0x17fdd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fdfc
ck-gc.c:mark:101:CK_DEBUG:0x7ffffb8: referenced start = 0x17fdfc
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fdfc, start=5, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fe34, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fe6c, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fea4, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fedc, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff14, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff4c, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff84, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ffbc, start=2, middle=0
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fff4, start=1, middle=0
TRACE:	GC:Checked 10 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
TRACE:GC: SUCCESS: no leaks!
ck-gc.c:mark_all:204:stack has 28 words
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fdfc, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fdfc
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fe34, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fe34
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fe6c, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fe6c
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fea4, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fea4
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fedc, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fedc
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff14, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17ff14
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff4c, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17ff4c
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ff84, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17ff84
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17ffbc, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17ffbc
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fff4, start=0, middle=0
TRACE:ERROR:GC:DEFINITE LEAK of 0x17fff4
TRACE:	nbytes=12
TRACE:	Block allocated at: tests/part2-test5.c:test:16
TRACE:	GC:Checked 10 blocks.
TRACE:		GC:ERRORS: 10 errors, 0 maybe_errors
----------------------------------------------------------
TRACE:GC: SUCCESS: found 10 leaks!
ck-gc.c:mark_all:204:stack has 26 words
---------------------------------------------------------
compacting:
TRACE:GC:FREEing ptr=0x17fdfc
ck-gc.c:sweep_free:283:freeing 0x17fdfc
found 0x17fdfc in 0x17fdd8
TRACE:GC:FREEing ptr=0x17fe34
ck-gc.c:sweep_free:283:freeing 0x17fe34
found 0x17fe34 in 0x17fe10
TRACE:GC:FREEing ptr=0x17fe6c
ck-gc.c:sweep_free:283:freeing 0x17fe6c
found 0x17fe6c in 0x17fe48
TRACE:GC:FREEing ptr=0x17fea4
ck-gc.c:sweep_free:283:freeing 0x17fea4
found 0x17fea4 in 0x17fe80
TRACE:GC:FREEing ptr=0x17fedc
ck-gc.c:sweep_free:283:freeing 0x17fedc
found 0x17fedc in 0x17feb8
TRACE:GC:FREEing ptr=0x17ff14
ck-gc.c:sweep_free:283:freeing 0x17ff14
found 0x17ff14 in 0x17fef0
TRACE:GC:FREEing ptr=0x17ff4c
ck-gc.c:sweep_free:283:freeing 0x17ff4c
found 0x17ff4c in 0x17ff28
TRACE:GC:FREEing ptr=0x17ff84
ck-gc.c:sweep_free:283:freeing 0x17ff84
found 0x17ff84 in 0x17ff60
TRACE:GC:FREEing ptr=0x17ffbc
ck-gc.c:sweep_free:283:freeing 0x17ffbc
found 0x17ffbc in 0x17ff98
TRACE:GC:FREEing ptr=0x17fff4
ck-gc.c:sweep_free:283:freeing 0x17fff4
found 0x17fff4 in 0x17ffd0
TRACE:	GC:Checked 10 blocks, freed 10, 120 bytes
compacted = 120
ck-gc.c:mark_all:204:stack has 28 words
---------------------------------------------------------
checking for leaks:
TRACE:	GC:Checked 0 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
TRACE:GC: SUCCESS: no leaks!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
