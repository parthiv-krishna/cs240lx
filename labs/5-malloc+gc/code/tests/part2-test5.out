find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./part2-test5.bin> about to boot
******************sending 10576 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
GC test: checking that a doubly-linked list is handled.
sizeof hdr=56, redzone=128
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1000b8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100200
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100348
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100490
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1005d8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100720
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100868
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x1009b0
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100af8
TRACE:allocating 12 bytes
TRACE:ckalloc:allocated 12 bytes, (total=328), ptr=0x100c40
TRACE:going to check heap: 3280 bytes allocated, 1045296 bytes left
TRACE:SUCCESS: checked 10 blocks, detected no errors
TRACE:GC: SUCCESS heap checked out
ck-gc.c:mark_all:216:stack has 38 words
---------------------------------------------------------
checking for leaks:
TRACE:	GC:Checked 10 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
TRACE:GC: SUCCESS: no leaks!
TRACE:going to check heap: 3280 bytes allocated, 1045296 bytes left
TRACE:SUCCESS: checked 10 blocks, detected no errors
TRACE:GC: SUCCESS heap checked out
ck-gc.c:mark_all:216:stack has 32 words
---------------------------------------------------------
checking for leaks:
TRACE:ERROR:GC:DEFINITE LEAK of 0x1000b8
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100200
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100348
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100490
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x1005d8
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100720
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100868
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x1009b0
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100af8
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:ERROR:GC:DEFINITE LEAK of 0x100c40
TRACE:	nbytes=12
TRACE:	Block allocated at: part2-test5.c:test:16
TRACE:	GC:Checked 10 blocks.
TRACE:		GC:ERRORS: 10 errors, 0 maybe_errors
----------------------------------------------------------
TRACE:GC: SUCCESS: found 10 leaks!
ck-gc.c:mark_all:216:stack has 30 words
---------------------------------------------------------
compacting:
TRACE:GC:FREEing ptr=0x1000b8
TRACE:freeing 0x1000b8
TRACE:GC:FREEing ptr=0x100200
TRACE:freeing 0x100200
TRACE:GC:FREEing ptr=0x100348
TRACE:freeing 0x100348
TRACE:GC:FREEing ptr=0x100490
TRACE:freeing 0x100490
TRACE:GC:FREEing ptr=0x1005d8
TRACE:freeing 0x1005d8
TRACE:GC:FREEing ptr=0x100720
TRACE:freeing 0x100720
TRACE:GC:FREEing ptr=0x100868
TRACE:freeing 0x100868
TRACE:GC:FREEing ptr=0x1009b0
TRACE:freeing 0x1009b0
TRACE:GC:FREEing ptr=0x100af8
TRACE:freeing 0x100af8
TRACE:GC:FREEing ptr=0x100c40
TRACE:freeing 0x100c40
TRACE:	GC:Checked 10 blocks, freed 10, 120 bytes
TRACE:	GC:total allocated = 120, total deallocated = 120
----------------------------------------------------------
compacted = 120
TRACE:going to check heap: 3280 bytes allocated, 1045296 bytes left
TRACE:SUCCESS: checked 10 blocks, detected no errors
TRACE:GC: SUCCESS heap checked out
ck-gc.c:mark_all:216:stack has 32 words
---------------------------------------------------------
checking for leaks:
TRACE:	GC:Checked 10 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
TRACE:GC: SUCCESS: no leaks!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
